File: Neetcode 150/Stack/evaluate-reverse-polish-notation.cpp

// https://leetcode.com/problems/evaluate-reverse-polish-notation/description/

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        
        stack<int> st;
        if(tokens.size() == 0) return stoi(tokens[0]);

        unordered_set<string> operands({"+", "-", "*", "/"});
        for(int i = 0; i < tokens.size(); i++){
            if(operands.find(tokens[i]) != operands.end()){
                int a = st.top();
                st.pop();
                int b = st.top();
                st.pop();
                int res = 0;
                if(tokens[i] == "+") res = a + b;
                else if(tokens[i] == "-") res = b-a;
                else if(tokens[i] == "*") res = a*b;
                else res = b/a;
                st.push(res);
            }else st.push(stoi(tokens[i]));
        }

        return st.top();
    }
};

File: Neetcode-150-js/Stack/evaluate-reverse-polish-notation.js

// https://leetcode.com/problems/evaluate-reverse-polish-notation/description/

/**
 * @param {string[]} tokens
 * @return {number}
 */

const OPERATORS = ['+', '-', '*', '/'];

var evalRPN = function(tokens) {
    let len = tokens.length;
    let result = 0;
    if(len === 0) return result;

    const stack = [];
    for(let i = 0; i < len; i++){
        let val = tokens[i];
        if(OPERATORS.includes(val)){
            let a = stack.pop();
            let b = stack.pop();

            if(val === '+') result = a+b;
            if(val === '-') result = b-a;
            if(val === '*') result = a*b;
            if(val === '/') result = Math.trunc(b/a);

            stack.push(result);
        }else stack.push(+val);
    }


    return stack.pop();
};